some enhancements for arc qemu

IRISZZW opened the pull request:
- enable MPU
- fix cpu frequency to 10Mhz

IRISZZW committed:
arc: qemu: enable MPU

Enable MPU for arc qemu.

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>

evgeniy-paltsev commented on soc/arc/snps_qemu/Kconfig.defconfig:
```diff
@@ -27,6 +27,9 @@ config NUM_IRQ_PRIO_LEVELS
 	default 25
 
+config ARC_MPU_VER
+	default 2
```
Do ARC HS processors can be configured with MPU version 2?

IRISZZW replied:
actually, it is version 3, but we don't support version 3 currently. and v3 is similar to v2 with minimal 32byte region size.
relative discussion:
https://github.com/zephyrproject-rtos/zephyr/issues/31416
https://github.com/zephyrproject-rtos/zephyr/pull/32573

abrodkin replied:
Right, so it worth addition of a comment explaining that. At least in the commit comment. Something like:
> Technically ARC HS supports MPUv3, but not v2. But given MPUv3 is the same as v2 but with minimal region size of 32 bytes, we may assume MPUv3 is just a subset of MPUv2.

IRISZZW commented on boards/arc/qemu_arc/board.cmake:
```diff
@@ -20,9 +20,11 @@ list(APPEND QEMU_FLAGS_${ARCH}
   -global cpu.num-irq=25
   -global cpu.ext-irq=20
-  -global cpu.freq_hz=1000000
+  -global cpu.freq_hz=50000000
```
I have explained in commit message when enabled MPU, some tests will execute timeout, then I tried to increase the frequency, then all tests passed. 
About `QEMU_ICOUNT_SHIFT`, I believe we need to do some changes, thanks for your reminding.
https://github.com/zephyrproject-rtos/zephyr/blob/7162786f83f960d40bbb17716246e2dc0e0690e9/boards/Kconfig#L40-L45

abrodkin commented on boards/arc/qemu_arc/qemu_arc_hs_defconfig:
```diff
@@ -12,3 +12,4 @@ CONFIG_UART_CONSOLE=y
 CONFIG_SERIAL=y
 CONFIG_UART_NS16550=y
 CONFIG_QEMU_ICOUNT_SHIFT=6
+CONFIG_ARC_MPU_ENABLE=y
```
Agree. Or somehow execute tests 2 times: w/ & w/o MPU enabled. Not sure though if that's easy doable. @nashif what do you think?

evgeniy-paltsev replied:
@abrodkin I guess It's easier not to reinvent the wheel and simply have two separate boards :)

abrodkin commented on boards/arc/qemu_arc/arc_mpu_regions.c:
```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2017 Synopsys
```
2021

ruuddw commented on soc/arc/snps_qemu/Kconfig.defconfig:
```diff
@@ -9,7 +9,7 @@ config SOC
 
 config SYS_CLOCK_HW_CYCLES_PER_SEC
 	int
-	default 10000000
```
50 Mhz is probably ok, but wasn't this the real problem for the 'timeouts'? QEMU configured for 100_000 and Zephyr configured for 1_000_000 cycles/sec?

IRISZZW replied:
I have tried to fix QEMU configuration and Zephyr configuration both to 10Mhz, but test: tests/kernel/mem_protect/syscalls/
still timeout. execution time:

- 10Mhz: 

```
real    2m1.306s
user    2m1.143s
sys     0m0.459s
```
- 50Mhz:
```
real    0m4.991s
user    0m4.869s
sys     0m0.485s
```

ruuddw replied:
Do you understand why this is? With iCount on (and the right shift values), it shouldn't matter that much. Maybe at 1Mhz the tick counter overhead too high?  (e.g. 1000 ticks/sec means only 1000 cycles/instruction for the handler).
Just increasing numbers without understanding why it is needed (and why 50 is  ok, instead of 10 or 100...) worries me a bit.

IRISZZW replied:
I can't reproduce it now, 10MHz works fine too.

ruuddw replied:
ok - please change the PR first message then to not mention changing to 50 MHz, but from 1 to 10 MHz. Any QEMU_ICOUNT_SHIFT to be updated as well?

IRISZZW replied:
we intend to use 10MHz, so the QEMU_ICOUNT_SHIFT is right.

IRISZZW committed:
board: qemu_arc: fix cpu frequency to 10Mhz

we have set SYS_CLOCK_HW_CYCLES_PER_SEC to 10000000,
so we need to set cpu.freq_hz=10000000 too.

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>

abrodkin commented:
@IRISZZW hm, see some tests still fail:
```
tests/kernel/fatal/exception/kernel.common.stack_protection in qemu_arc_em:tests/kernel/fatal/exception/kernel.common
tests/kernel/mem_protect/mem_protect/kernel.memory_protection in qemu_arc_em:tests/kernel/mem_protect/mem_protect/kernel.memory_protection
tests/kernel/fatal/exception/kernel.common.stack_protection in qemu_arc_hs:tests/kernel/fatal/exception/kernel.common
tests/kernel/mem_protect/mem_protect/kernel.memory_protection in qemu_arc_hs:tests/kernel/mem_protect/mem_protect/kernel.memory_protection
```

IRISZZW commented:
> @IRISZZW hm, see some tests still fail:
> 
> ```
> tests/kernel/fatal/exception/kernel.common.stack_protection in qemu_arc_em:tests/kernel/fatal/exception/kernel.common
> tests/kernel/mem_protect/mem_protect/kernel.memory_protection in qemu_arc_em:tests/kernel/mem_protect/mem_protect/kernel.memory_protection
> tests/kernel/fatal/exception/kernel.common.stack_protection in qemu_arc_hs:tests/kernel/fatal/exception/kernel.common
> tests/kernel/mem_protect/mem_protect/kernel.memory_protection in qemu_arc_hs:tests/kernel/mem_protect/mem_protect/kernel.memory_protection
> ```

`tests/kernel/fatal/exception/kernel.common.stack_protection`  issue caused by wrong configuration of `CONFIG_ARCH_HAS_STACK_PROTECTION`, I have created a pull request (#32967) to fix it.
`tests/kernel/mem_protect/mem_protect/kernel.memory_protection` issue caused by some bugs of qemu-arc, relative issue:
[No privilege exception raise when execute clri in user mode](https://github.com/foss-for-synopsys-dwc-arc-processors/qemu/issues/14)

abrodkin commented:
So I guess to get clean results we need to black-list for now `tests/kernel/mem_protect/mem_protect/kernel.memory_protection` until https://github.com/foss-for-synopsys-dwc-arc-processors/qemu/issues/14 is fixed and fix gets propagated to new SDK.

IRISZZW committed:
board: qemu_arc: disable test: tests/kernel/mem_protect/mem_protect

This test fails on qemu_arc_{em|hs} consistently,
due to bug in quem_arc, details:
https://github.com/foss-for-synopsys-dwc-arc-processors/qemu/issues/14.
To get clean results we need to disable this test until the
bug is fixed and fix gets propagated to new Zephyr-SDK.

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>

IRISZZW commented:
@nashif Could you please help to get it merged?

IRISZZW commented:
buildkit failures not caused by this PR.

```
8 failures:

tests/net/igmp/net.igmp in frdm_k64f:tests/net/igmp/net.igmp
tests/net/igmp/net.igmp.preempt in frdm_k64f:tests/net/igmp/net.igmp
tests/net/igmp/net.igmp in native_posix:tests/net/igmp/net.igmp
tests/net/igmp/net.igmp.preempt in native_posix:tests/net/igmp/net.igmp
tests/net/igmp/net.igmp in qemu_x86:tests/net/igmp/net.igmp
tests/net/igmp/net.igmp.preempt in qemu_x86:tests/net/igmp/net.igmp
tests/net/igmp/net.igmp in mps2_an385:tests/net/igmp/net.igmp
tests/net/igmp/net.igmp.preempt in mps2_an385:tests/net/igmp/net.igmp
```

abrodkin commented:
> buildkit failures not caused by this PR.
> 
> ```
> 8 failures:
> 
> tests/net/igmp/net.igmp in frdm_k64f:tests/net/igmp/net.igmp
> tests/net/igmp/net.igmp.preempt in frdm_k64f:tests/net/igmp/net.igmp
> tests/net/igmp/net.igmp in native_posix:tests/net/igmp/net.igmp
> tests/net/igmp/net.igmp.preempt in native_posix:tests/net/igmp/net.igmp
> tests/net/igmp/net.igmp in qemu_x86:tests/net/igmp/net.igmp
> tests/net/igmp/net.igmp.preempt in qemu_x86:tests/net/igmp/net.igmp
> tests/net/igmp/net.igmp in mps2_an385:tests/net/igmp/net.igmp
> tests/net/igmp/net.igmp.preempt in mps2_an385:tests/net/igmp/net.igmp
> ```

Indeed this is now fixed with https://github.com/zephyrproject-rtos/zephyr/pull/34704 I guess.

galak commented:
@IRISZZW can you take a look at the new CI failures.

IRISZZW committed:
board: qemu_arc: skip trigger_fault_divide_zero test case

qemu arc has limited support for DIV/REM, no DIVZERO exception
throw, so we need to skip trigger_fault_divide_zero test case.

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>
